spec/Candid.md

```diff
@@ -1188,11 +1200,15 @@ M : <val> -> <reftype> -> i8*
 M(ref(r) : service <actortype>) = i8(0)
 M(id(v*) : service <actortype>) = i8(1) M(v* : vec nat8)
 
-M(ref(r)   : func <functype>) = i8(0)
-M(pub(s,n) : func <functype>) = i8(1) M(s : service {}) M(n : text)
+M(ref(r)         : func <functype>) = i8(0)
+M(pub(s,n)       : func <functype>) = i8(1) M(s : service {}) M(n : text)
+M(pub(s,n,v*:t*) : func <functype>) = i8(2) M(s : service {}) M(n : text) M*(v* : t*)
 
 M(ref(r) : principal) = i8(0)
 M(id(v*) : principal) = i8(1) M(v* : vec nat8)
+
+M* : <val>* -> <datatype>* -> i8*
+M*(v^N : <datatype>^N) = leb128(N) M(v : <datatype>)^N
```

Comment:
this will be a breaking change? or the bound argument is stored in `R` which was not used before?

Reply:
No, R is only used if there are reference types that need it, as before. The type is determined from M.

But I think this is a breaking change in so far that a receiver not understanding the new encoding of closures would choke on it. The only way to avoid this is by indeed introducing a separate closure type as a future type, it seems. Hm, that would be ugly...

Reply:
A new type seems to be appropriate. Not breaking existing stuff is part of our value proposition, and given the complexity of this maybe it’s good if a service can say that they really only support plain function references, but not closures?